Repository/ContainerRepositoryFactory.php

```diff
@@ -39,7 +39,7 @@ public function __construct(ContainerInterface $container = null)
     /**
      * {@inheritdoc}
      */
-    public function getRepository(EntityManagerInterface $entityManager, $entityName)
+    public function getRepository(EntityManagerInterface $entityManager, $entityName) : ObjectRepository
```

Comment:
This change is in fact a BC break because before repositories were not enforced to be subclass of ObjectRepository.

Reply:
But `ContainerRepositoryFactory` is final and users can't extends from it. Isn't it?

Reply:
> But `ContainerRepositoryFactory` is final and users can't extends from it. Isn't it?

Precisely.

Reply:
@andrew-demb Yes but in Symfony ContainerRepositoryFactory overrides default RepositoryFactory from Doctrine, making it a BC break. Not a BC break in the contract of this class but a BC break in the bundle itself. It's no longer possible to use custom repositories not extending ObjectRepository and doing so makes the app crash due to this added typehint.

Reply:
It's been documented as supposed to be `ObjectRepository` since the creation of `RepositoryFactory` 7 years ago (!) : https://github.com/doctrine/orm/commit/7eb744126ba5081ea51441a893dd950046777383

repositories were not enforced to be subclass of ObjectRepository because type declarations did not exist back then, but even if the contract is not enforced, it's still a contract. Call me heartless, but I can't say I feel sorry for apps experiencing this crash, they kind of had it coming, don't you agree?

Reply:
I don't agree, this kind of stuff does not belong to a patch version.

Reply:
Requirement for the ObjectRepository/EntityRepository was never documented: https://github.com/doctrine/orm/blob/v2.6.0/lib/Doctrine/ORM/EntityManager.php#L712
It was also never validated: https://github.com/doctrine/orm/blob/v2.7.1/docs/en/reference/working-with-objects.rst#by-native-queries
There is return annotation for EntityManager::getRepository(): https://github.com/doctrine/orm/blob/v2.6.0/lib/Doctrine/ORM/EntityManager.php#L712 but that is the only public API and has never been enforced anyway.

It's not nice "feature" and I don't like it either but it was working and documented behaviour (_" You can overwrite this behaviour by specifying the class name of your own Entity Repository in the Annotation, XML or YAML metadata."_).

Reply:
> I don't personally need this "feature" but I don't think this should've been merged as is.

You are, as always, entitled to your opinion. To look at the facts:
`ContainerRepositoryFactory` has always implemented `RepositoryFactory` from ORM with an `@inheritdoc` docblock.
`RepositoryFactory` in ORM has had an `@return ObjectRepository` on its `getRepository` method since this commit in 2013, where it was widened from ORM's `EntityRepository` to the persistence library's `ObjectRepository`: https://github.com/doctrine/orm/commit/3488049c18849a0cb942f4fea00501fa11727a83.

I don't know where you see the BC break, but if we treat docblocks as a contract (which we always have and always should), this change is perfectly fine. You are always welcome to improve documentation which is clearly wrong in this case, as the code contains the absolute truth, which is that the contract always specified an `ObjectRepository` return type for as long as the class has existed.

Reply:
> I don't know where you see the BC break

The BC break I see is that upgrading doctrine/doctrine-bundle from 2.0.6 to 2.0.7 leads to fatal errors.

> as the code contains the absolute truth

The code can actually contain bugs.

Reply:
> The BC break I see is that upgrading doctrine/doctrine-bundle from 2.0.6 to 2.0.7 leads to fatal errors.

Please post an example that passes on 2.0.6 and produces a fatal error on 2.0.7 and I'll fix it. Thanks!